pgmpy/factors/Factor.py

```diff
@@ -780,10 +780,12 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
     def __hash__(self):
-        variables = sorted(self.variables)
+        variable_hashes = [hash(variable) for variable in self.variables]
+        sorted_var_hashes = list(sorted(variable_hashes))
         phi = self.copy()
         for axis in range(phi.values.ndim):
-            exchange_index = phi.variables.index(variables[axis])
+            exchange_index = variable_hashes.index(sorted_var_hashes[axis])
+            variable_hashes[axis], variable_hashes[exchange_index] = (variable_hashes[exchange_index], variable_hashes[axis])
             phi.variables[axis], phi.variables[exchange_index] = (phi.variables[exchange_index],
```

Comment:
@khalibartan Do you want me to replace 
`hash(str(phi.variables) + str(phi.values) + str(phi.cardinality))`
with
`hash(sorted_var_hashes+ str(phi.values) + str(phi.cardinality))` ?
I will clean the commit history also.